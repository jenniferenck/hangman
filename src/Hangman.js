import React, { Component } from 'react';
import { randomWord } from './words';
import './Hangman.css';
import img0 from './0.jpg';
import img1 from './1.jpg';
import img2 from './2.jpg';
import img3 from './3.jpg';
import img4 from './4.jpg';
import img5 from './5.jpg';
import img6 from './6.jpg';

class Hangman extends Component {
  /** by default, allow 6 guesses and use provided gallows images. */
  static defaultProps = {
    maxWrong: 6,
    images: [img0, img1, img2, img3, img4, img5, img6]
  };

  constructor(props) {
    super(props);
    this.state = { nWrong: 0, guessed: new Set(), answer: randomWord() };
    this.handleGuess = this.handleGuess.bind(this);
    this.printrandomWord = this.printrandomWord.bind(this);
    this.restartGame = this.restartGame.bind(this);
  }

  printrandomWord() {
    console.log(this.state.answer);
    return this.state.answer;
  }
  /** guessedWord: show current-state of word:
    if guessed letters are {a,p,e}, show "app_e" for "apple"
  */
  guessedWord() {
    return this.state.answer
      .split('')
      .map(ltr => (this.state.guessed.has(ltr) ? ltr : '_'));
  }

  /** handleGuest: handle a guessed letter:
    - add to guessed letters
    - if not in answer, increase number-wrong guesses
  */
  handleGuess(evt) {
    let ltr = evt.target.value;
    this.setState(state => {
      // make a copy of the set
      const newGuessed = new Set([...state.guessed.values(), ltr]);

      // tally number wrong
      let newNWrong = state.nWrong;
      if (!state.answer.includes(ltr)) {
        newNWrong += 1;
      }

      // finally udate the state
      return {
        guessed: newGuessed,
        nWrong: newNWrong
      };
    });
  }

  restartGame() {
    // on click, set state of game back to original and get new random word
    this.setState({ nWrong: 0, guessed: new Set(), answer: randomWord() });
  }

  /** generateButtons: return array of letter buttons to render */
  generateButtons() {
    return 'abcdefghijklmnopqrstuvwxyz'.split('').map(ltr => (
      <button
        className="keyboard"
        key={ltr}
        value={ltr}
        u
        onClick={this.handleGuess}
        disabled={
          this.state.nWrong < this.props.maxWrong
            ? this.state.guessed.has(ltr)
            : true
        }
      >
        {ltr}
      </button>
    ));
  }

  /** render: render game */
  render() {
    return (
      <div className="Hangman">
        <img src={this.props.images[this.state.nWrong]} />
        <p className="Hangman-word">{this.guessedWord()}</p>
        <div>{this.generateButtons()}</div>
        {this.state.nWrong === this.props.maxWrong ? (
          <p>GAME OVER SUCKER the answer is {this.state.answer}</p>
        ) : (
          <p>You guessed {this.state.nWrong} wrong</p>
        )}
        <button onClick={this.restartGame}>Restart</button>
      </div>
    );
  }
}

export default Hangman;
