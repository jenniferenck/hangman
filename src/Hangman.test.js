import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Hangman from './Hangman';
import ReactDOM from 'react-dom';

describe('<Hangman />', function() {
  // smoke test
  it('renders without crashing', function() {
    shallow(<Hangman />);
  });

  // snapshot test
  it('matches snapshot', function() {
    const wrapper = shallow(<Hangman />);
    const serialized = toJson(wrapper);
    expect(serialized).toMatchSnapshot();
  });
});

/* enzyme selector tests */

// default state
it('has a default STATES of nWrong and answer to start', function() {
  const wrapper = shallow(<Hangman />);
  // has two default states: images and maxWrong
  expect(wrapper.state().nWrong).toBe(0);
  // call the printanswer fuc
  const answer = wrapper.state().answer;
  expect(wrapper.state().answer).toBe(answer);
  expect(wrapper.state().guessed).toBeInstanceOf(Set);
});

// default props
it('has a default PROPS of nWrong and answer to start', function() {
  const wrapper = mount(<Hangman />);

  expect(wrapper.props()).toHaveProperty('maxWrong');
  expect(wrapper.props().maxWrong).toBe(6);
});

// generateButton()
it('generates a button for each letter', function() {
  const wrapper = shallow(<Hangman />);

  const button = wrapper.find('.keyboard');
  //   make sure 26 keys were generated
  expect(button).toHaveLength(26);
});

// test single button works
it('test single disables after click', function() {
  const wrapper = mount(<Hangman />);
  expect(wrapper.state().guessed.has('a')).toEqual(false);

  //   simulate click on 'A' button and pass in simulated evt target
  let buttonA = wrapper.find("button[value='a']");
  buttonA.simulate('click', { target: { value: 'a' } });
  buttonA = wrapper.find("button[value='a']");

  //   expect(wrapper.state().guessed.has('a')).toEqual(true);
  expect(buttonA.props().disabled).toBe(true);
});

// test that game ends when nWrong > maxWrong
it('test that all buttons disable at end of game', function() {
  const wrapper = shallow(<Hangman />);
  wrapper.setState({ nWrong: 7 });
  console.log(wrapper.state());
  // all buttons should be disabled
  const buttons = wrapper.find('.keyboard');
  buttons.forEach(b => {
    expect(b.props().disabled).toBe(true);
  });
});

// ADD TEST that RESTART BUTTON WORKS
