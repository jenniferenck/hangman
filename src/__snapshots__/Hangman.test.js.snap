// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Hangman /> matches snapshot 1`] = `
<div
  className="Hangman"
>
  <img
    src="0.jpg"
  />
  <p
    className="Hangman-word"
  >
    _
    _
    _
    _
    _
  </p>
  <div>
    <button
      className="keyboard"
      disabled={false}
      key="a"
      onClick={[Function]}
      u={true}
      value="a"
    >
      a
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="b"
      onClick={[Function]}
      u={true}
      value="b"
    >
      b
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="c"
      onClick={[Function]}
      u={true}
      value="c"
    >
      c
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="d"
      onClick={[Function]}
      u={true}
      value="d"
    >
      d
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="e"
      onClick={[Function]}
      u={true}
      value="e"
    >
      e
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="f"
      onClick={[Function]}
      u={true}
      value="f"
    >
      f
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="g"
      onClick={[Function]}
      u={true}
      value="g"
    >
      g
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="h"
      onClick={[Function]}
      u={true}
      value="h"
    >
      h
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="i"
      onClick={[Function]}
      u={true}
      value="i"
    >
      i
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="j"
      onClick={[Function]}
      u={true}
      value="j"
    >
      j
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="k"
      onClick={[Function]}
      u={true}
      value="k"
    >
      k
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="l"
      onClick={[Function]}
      u={true}
      value="l"
    >
      l
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="m"
      onClick={[Function]}
      u={true}
      value="m"
    >
      m
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="n"
      onClick={[Function]}
      u={true}
      value="n"
    >
      n
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="o"
      onClick={[Function]}
      u={true}
      value="o"
    >
      o
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="p"
      onClick={[Function]}
      u={true}
      value="p"
    >
      p
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="q"
      onClick={[Function]}
      u={true}
      value="q"
    >
      q
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="r"
      onClick={[Function]}
      u={true}
      value="r"
    >
      r
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="s"
      onClick={[Function]}
      u={true}
      value="s"
    >
      s
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="t"
      onClick={[Function]}
      u={true}
      value="t"
    >
      t
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="u"
      onClick={[Function]}
      u={true}
      value="u"
    >
      u
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="v"
      onClick={[Function]}
      u={true}
      value="v"
    >
      v
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="w"
      onClick={[Function]}
      u={true}
      value="w"
    >
      w
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="x"
      onClick={[Function]}
      u={true}
      value="x"
    >
      x
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="y"
      onClick={[Function]}
      u={true}
      value="y"
    >
      y
    </button>
    <button
      className="keyboard"
      disabled={false}
      key="z"
      onClick={[Function]}
      u={true}
      value="z"
    >
      z
    </button>
  </div>
  <p>
    You guessed 
    0
     wrong
  </p>
  <button
    onClick={[Function]}
  >
    Restart
  </button>
</div>
`;
